### DISCUSSION TITLE
Wix 5 extension usage from command line only

### DISCUSSION THREAD

**Discussion by ijuedt** [wixtoolset/issues#8669] [Questions]
In Wix 3 it was possible to use the Wix extension dlls directly by placing them in the wix binary directory and then declaring their usage in the command line (e. g. "-ext WixToolset.Netfx.wixext"). This possibility seems to have gone in Wix 4/5. I am migrating a build job operating on command line only without using a project or an extension cache (so not using nupkg either). I also tried to install the extensions by pushing the add command to command line ("wix extension add -g WixToolset.Netfx.wixext"), but this results in an error "Could not find expected package root folder wixext5". Of course, the "add" command works manually, but I need it as part of a command line called from a Python shell script that will run on a build machine. With Wix 3, that worked fine. How this is supposed to work in Wix 4/5?

**Comment by robmen** [comment]
I ran the following commands, and things looked good:
```cmd
C:\> wix --version
5.0.1+2f00cbe6
C:\> wix extension add -g WixToolset.Netfx.wixext/5.0.1
C:\> wix extension list -g
WixToolset.Netfx.wixext 5.0.1
```

Note: It's a good idea to specify the version when installing an extension to ensure it is compatible with your version of the WiX toolset. So you should do the following instead.

```cmd
C:\> wix extension add -g WixToolset.Netfx.wixext/5.0.1
```

However, I recommend a different approach. Using a .wixproj and `dotnet build` is much easier than the `wix.exe` command line.

For example: my.wixproj
```xml
<Project Sdk="WixToolset.Sdk/5.0.1">
   <PackageReference Include="WixToolset.Netfx.wixext" Version="5.0.1" />
</Project>
```
and my.wxs:
```xml
<Wix ...>
...
</Wix>
```

Then, to build:
```cmd
dotnet build
```

I find two files and one command line easier to manage than two command lines and one file. Plus, you're better set up for change. For example, it is easy to add more .wxs files, and you don't need to change the command line or .wixproj file.

**Reply by ijuedt** [reply]
It is no problem for me to run these commands directly. But if I cobble them together in Python with "cmd.append('wix extension add -g {0}/5.0.1'.format(extension))" (with or without the exact version) and invoke with "subprocess.run", then running the script results in error "wix : warning WIX6101: Could not find expected package root folder wixext5. Ensure wix/5.0.1 is compatible with WiX v5."

The same script will run the "build" command correctly, so I hoped I could do the same with "extension add".

But am I right in understanding that the Wix 3 feature to just put extension dlls into the wix.exe binary directory has been dropped?

Edit: it is not feasable for me (at least in the short run) to migrate to a wixproj, as this would imply a major rewrite.

**Reply by bevanweiss** [reply]
> Edit: it is not feasable for me (at least in the short run) to migrate to a wixproj, as this would imply a major rewrite.

Can you explain why this would be the case?
You currently have a collection of wxs/wxi files right?
Creating a .wixproj file that references them should not be a rewrite of anything.
The .wixproj really just rolls up a bunch of the command line options you're throwing in, and puts them in a project file.
.wixproj does not change your original wxs/wxi authoring.

**Reply by robmen** [reply]
Yeah, the .wixproj doesn't even explicitly reference the source files as long as they are all in the same directory.

Otherwise, @ijuedt, it sounds like something in your Python script is messing stuff up. It is hard to say what is unique about your environment. I don't use Python to drive my build processes.

**Reply by ijuedt** [reply]
> > Edit: it is not feasable for me (at least in the short run) to migrate to a wixproj, as this would imply a major rewrite.
> 
> Can you explain why this would be the case? You currently have a collection of wxs/wxi files right? Creating a .wixproj file that references them should not be a rewrite of anything. The .wixproj really just rolls up a bunch of the command line options you're throwing in, and puts them in a project file. .wixproj does not change your original wxs/wxi authoring.

@bevanweiss: I have a legacy project that relies heavily on python scripting and is integrated into an enterprise environment, with shared libraries responsible for more than one product. Wix is used by the command line only. If a wixproj can be integrated in such an environment, however, I could at least try a proof-of-concept.

**Reply by ijuedt** [reply]
> Yeah, the .wixproj doesn't even explicitly reference the source files as long as they are all in the same directory.
> 
> Otherwise, @ijuedt, it sounds like something in your Python script is messing stuff up. It is hard to say what is unique about your environment. I don't use Python to drive my build processes.

@robmen: I'm looking exactly for that: where the script might be messing up. :-) I hoped to learn something about why the script has been working fine with Wix 3, but fails to do so with Wix 5.
